File: sdk/node/examples/streaming/hitl_confirmation_loop.ts
#!/usr/bin/env npx tsx
/**
 * Minimal streaming HITL confirmation loop.
 *
 * Usage:
 *   A3S_CONFIG_FILE=./agent.acl npx tsx streaming/hitl_confirmation_loop.ts
 *   A3S_CONFIG_FILE=./agent.acl A3S_CODE_HITL_AUTO_APPROVE=1 npx tsx streaming/hitl_confirmation_loop.ts
 */

import { createInterface } from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';
import { Agent } from '../../index.js';
import type { AgentEvent, PendingConfirmation, Session, SessionOptions } from '@a3s-lab/code';

type ConfirmationEventData = {
  args?: unknown;
  timeout_ms?: number;
};

function eventData(event: AgentEvent): ConfirmationEventData {
  if (!event.data) return {};
  try {
    return JSON.parse(event.data) as ConfirmationEventData;
  } catch {
    return {};
  }
}

async function pendingByToolId(session: Session, toolId?: string | null): Promise<PendingConfirmation | null> {
  const pending = await session.pendingConfirmations();
  return pending.find(item => item.toolId === toolId) ?? pending[0] ?? null;
}

async function confirmFromCli(
  session: Session,
  event: AgentEvent,
  ask: (question: string) => Promise<string>,
): Promise<void> {
  const pending = await pendingByToolId(session, event.toolId);
  const data = eventData(event);
  const toolId = pending?.toolId ?? event.toolId;

  if (!toolId) {
    console.warn('[hitl] confirmation_required event has no tool id; skipping');
    return;
  }

  const toolName = pending?.toolName ?? event.toolName ?? 'unknown';
  const args = pending?.args ?? data.args ?? {};
  const remainingMs = pending?.remainingMs ?? data.timeout_ms;

  console.log('\n[hitl] Tool requires confirmation');
  console.log(`tool: ${toolName}`);
  console.log(`toolId: ${toolId}`);
  if (remainingMs !== undefined) console.log(`remainingMs: ${remainingMs}`);
  console.log(JSON.stringify(args, null, 2));

  const autoApprove = process.env.A3S_CODE_HITL_AUTO_APPROVE === '1';
  const answer = autoApprove ? 'y' : await ask('Approve this tool call? [y/N] ');
  const approved = answer.trim().toLowerCase().startsWith('y');

  const completed = await session.confirmToolUse(
    toolId,
    approved,
    approved ? 'Approved from Node HITL example' : 'Rejected from Node HITL example',
  );
  console.log(`[hitl] confirmation ${completed ? 'resolved' : 'not found'} (${approved ? 'approved' : 'rejected'})`);
}

async function main(): Promise<void> {
  const configPath = process.env.A3S_CONFIG_FILE ?? process.argv[2];
  if (!configPath) {
    throw new Error('Set A3S_CONFIG_FILE or pass an agent.acl path as the first argument.');
  }

  const workspace = process.env.A3S_CODE_WORKSPACE ?? process.cwd();
  const prompt = process.env.A3S_CODE_HITL_PROMPT
    ?? 'Use the bash tool to run: echo A3S_CODE_HITL_OK. Then summarize the result.';

  const options: SessionOptions = {
    planningMode: 'disabled',
    permissionPolicy: {
      ask: ['bash*'],
      defaultDecision: 'allow',
    },
    confirmationPolicy: {
      enabled: true,
      defaultTimeoutMs: 120000,
      timeoutAction: 'reject',
      yoloLanes: ['query'],
    },
  };

  const agent = await Agent.create(configPath);
  const session = agent.session(workspace, options);
  const rl = createInterface({ input, output });

  try {
    const stream = await session.stream(prompt);

    while (true) {
      const next = await stream.next();
      if (next.done || !next.value) break;

      const event = next.value;
      if (event.type === 'text_delta' && event.text) {
        process.stdout.write(event.text);
      } else if (event.type === 'tool_start') {
        console.log(`\n[tool:start] ${event.toolName ?? 'unknown'}`);
      } else if (event.type === 'tool_end') {
        console.log(`\n[tool:end] ${event.toolName ?? 'unknown'} exit=${event.exitCode ?? 0}`);
      } else if (event.type === 'confirmation_required') {
        await confirmFromCli(session, event, question => rl.question(question));
      } else if (event.type === 'confirmation_received') {
        console.log(`\n[hitl] confirmation_received ${event.toolId ?? ''}`);
      } else if (event.type === 'confirmation_timeout') {
        console.log(`\n[hitl] confirmation_timeout ${event.toolId ?? ''}`);
      } else if (event.type === 'error') {
        throw new Error(event.error ?? 'stream error');
      }
    }

    console.log('\n[hitl] stream complete');
  } finally {
    rl.close();
    await session.cancelConfirmations();
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: sdk/python/examples/hitl_confirmation_loop.py
#!/usr/bin/env python3
"""Minimal streaming HITL confirmation loop.

Usage:
    A3S_CONFIG_FILE=./agent.acl python examples/hitl_confirmation_loop.py
    A3S_CONFIG_FILE=./agent.acl A3S_CODE_HITL_AUTO_APPROVE=1 python examples/hitl_confirmation_loop.py
"""

from __future__ import annotations

import json
import os
import sys
from typing import Any

from a3s_code import Agent, ConfirmationPolicy, PermissionPolicy, SessionOptions


def event_data(event: Any) -> dict[str, Any]:
    if not event.data:
        return {}
    try:
        value = json.loads(event.data)
        return value if isinstance(value, dict) else {}
    except json.JSONDecodeError:
        return {}


def pending_by_tool_id(session: Any, tool_id: str | None) -> dict[str, Any] | None:
    pending = session.pending_confirmations()
    for item in pending:
        if item.get("tool_id") == tool_id:
            return item
    return pending[0] if pending else None


def confirm_from_cli(session: Any, event: Any) -> None:
    pending = pending_by_tool_id(session, event.tool_id)
    data = event_data(event)
    tool_id = (pending or {}).get("tool_id") or event.tool_id

    if not tool_id:
        print("[hitl] confirmation_required event has no tool id; skipping", file=sys.stderr)
        return

    tool_name = (pending or {}).get("tool_name") or event.tool_name or "unknown"
    args = (pending or {}).get("args") or data.get("args") or {}
    remaining_ms = (pending or {}).get("remaining_ms") or data.get("timeout_ms")

    print("\n[hitl] Tool requires confirmation")
    print(f"tool: {tool_name}")
    print(f"tool_id: {tool_id}")
    if remaining_ms is not None:
        print(f"remaining_ms: {remaining_ms}")
    print(json.dumps(args, indent=2, ensure_ascii=False))

    auto_approve = os.environ.get("A3S_CODE_HITL_AUTO_APPROVE") == "1"
    answer = "y" if auto_approve else input("Approve this tool call? [y/N] ")
    approved = answer.strip().lower().startswith("y")
    completed = session.confirm_tool_use(
        tool_id,
        approved,
        "Approved from Python HITL example" if approved else "Rejected from Python HITL example",
    )
    state = "approved" if approved else "rejected"
    print(f"[hitl] confirmation {'resolved' if completed else 'not found'} ({state})")


def main() -> None:
    config_path = os.environ.get("A3S_CONFIG_FILE") or (sys.argv[1] if len(sys.argv) > 1 else None)
    if not config_path:
        raise RuntimeError("Set A3S_CONFIG_FILE or pass an agent.acl path as the first argument.")

    workspace = os.environ.get("A3S_CODE_WORKSPACE") or os.getcwd()
    prompt = os.environ.get("A3S_CODE_HITL_PROMPT") or (
        "Use the bash tool to run: echo A3S_CODE_HITL_OK. Then summarize the result."
    )

    opts = SessionOptions()
    opts.planning_mode = "disabled"
    opts.permission_policy = PermissionPolicy(ask=["bash*"], default_decision="allow")
    opts.confirmation_policy = ConfirmationPolicy(
        enabled=True,
        default_timeout_ms=120_000,
        timeout_action="reject",
        yolo_lanes=["query"],
    )

    agent = Agent.create(config_path)
    session = agent.session(workspace, opts)

    try:
        for event in session.stream(prompt):
            if event.event_type == "text_delta" and event.text:
                print(event.text, end="", flush=True)
            elif event.event_type == "tool_start":
                print(f"\n[tool:start] {event.tool_name or 'unknown'}")
            elif event.event_type == "tool_end":
                print(f"\n[tool:end] {event.tool_name or 'unknown'} exit={event.exit_code or 0}")
            elif event.event_type == "confirmation_required":
                confirm_from_cli(session, event)
            elif event.event_type == "confirmation_received":
                print(f"\n[hitl] confirmation_received {event.tool_id or ''}")
            elif event.event_type == "confirmation_timeout":
                print(f"\n[hitl] confirmation_timeout {event.tool_id or ''}")
            elif event.event_type == "error":
                raise RuntimeError(event.error or "stream error")

        print("\n[hitl] stream complete")
    finally:
        session.cancel_confirmations()


if __name__ == "__main__":
    main()
